File: ApiDemos/project/java-app/src/main/java/com/example/mapdemo/DataDrivenDatasetStylingActivity.java
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.example.mapdemo;

import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.graphics.ColorUtils;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

//noinspection UnusedImport
import com.example.common_ui.R; // <-- Keep this import

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMapOptions;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.DatasetFeature;
import com.google.android.gms.maps.model.Feature;
import com.google.android.gms.maps.model.FeatureClickEvent;
import com.google.android.gms.maps.model.FeatureLayer;
import com.google.android.gms.maps.model.FeatureLayerOptions;
import com.google.android.gms.maps.model.FeatureStyle;
import com.google.android.gms.maps.model.FeatureType;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MapCapabilities;

import java.util.List;
import java.util.Map;


/**
 * This sample showcases how to use the Data-driven styling for datasets. For more information
 * on how the Data-driven styling for boundaries work, check out the following link:
 * https://developers.google.com/maps/documentation/android-sdk/dds-datasets/overview
 * <p>
 * This is meant to work with the datasets in the res/raw directory.
 */
// [START maps_android_data_driven_styling_datasets]
public class DataDrivenDatasetStylingActivity extends SamplesBaseActivity implements OnMapReadyCallback, FeatureLayer.OnFeatureClickListener {
    private record DataSet(
            String label,
            String datasetId,
            LatLng location,
            float zoomLevel,
            DataDrivenDatasetStylingActivity.DataSet.StylingCallback callback) {
            public interface StylingCallback {
                void styleDatasetLayer();
            }
    }

    /**
     * An array of DataSet objects representing different geographic locations and their associated data.
     * Each DataSet contains:
     *  - A human-readable name (e.g., "Boulder", "New York").
     *  - A unique Dataset ID, which should correspond to a dataset id in the Datasets console tab.
     *  - The central latitude and longitude coordinates (LatLng) of the location.
     *  - A styling function (method reference) that defines how to style the data from that dataset on a map.
     * <p>
     * This array is used to configure which datasets are available for display and how they should be presented.
     * Each element of the array should be a new DataSet object.
     * Modify the constructor arguments of each DataSet to define your specific data and styles.
     * The styling function will receive a `Layer` to which it can add elements.
     * <p>
     * Example:
     *   - `new DataSet("Boulder", "Boulder-DataSet-Id", new LatLng(40.0150, -105.2705), this::styleBoulderDatasetLayer)`
     *     This creates a DataSet for Boulder, identified by "Boulder-DataSet-Id", centered on the given coordinates,
     *     and styled using the `styleBoulderDatasetLayer` method.
     * <p>
     * Note: We have use the secrets plugin to allow us to configure the Dataset IDs in our secrets.properties file.
     */
    private final DataSet[] dataSets = new DataSet[] {
            new DataSet("Boulder", BuildConfig.BOULDER_DATASET_ID, new LatLng(40.0150, -105.2705), 11f, this::styleBoulderDatasetLayer),
            new DataSet("New York", BuildConfig.NEW_YORK_DATASET_ID, new LatLng(40.786244, -73.962684), 14f, this::styleNYCDatasetLayer),
            new DataSet("Kyoto", BuildConfig.KYOTO_DATASET_ID, new LatLng(35.005081, 135.764385), 13.5f, this::styleKyotoDatasetsLayer),
    };

    private DataSet findDataSetByLabel(String label) {
        for (DataSet dataSet : dataSets) {
            if (dataSet.label().equalsIgnoreCase(label)) { // Case-insensitive comparison
                return dataSet;
            }
        }
        return null; // Return null if no match is found
    }

    private static final String TAG = DataDrivenDatasetStylingActivity.class.getName();
    private static FeatureLayer datasetLayer = null;
    private GoogleMap map;
    private String lastGlobalId = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.data_driven_styling_demo);

        // [START_EXCLUDE silent]
        // 1. Get the Application instance and cast it
        ApiDemoApplication app = (ApiDemoApplication) getApplication();

        // 2. Call the getMapId() method
        String mapId = app.getMapId();

        if (mapId == null) {
            finish();
            return;
        }
        // [END_EXCLUDE]

        // --- Programmatically Create and Add Map Fragment ---
        // 1. Create GoogleMapOptions
        GoogleMapOptions mapOptions = new GoogleMapOptions();

        // 2. Set the mapId from the secrets.properties file
        mapOptions.mapId(BuildConfig.MAP_ID); // Use the mapId retrieved earlier

        // 3. Create SupportMapFragment instance with options
        SupportMapFragment mapFragment = SupportMapFragment.newInstance(mapOptions);
        mapFragment.getMapAsync(this);

        // 4. Add the fragment to your FrameLayout container using FragmentManager
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.map_fragment_container, mapFragment); // Use the container ID from XML
        fragmentTransaction.commit();
        // --- End Programmatic Creation ---


        int[] buttonIds = {R.id.button_boulder, R.id.button_ny, R.id.button_kyoto};
        for (int buttonId : buttonIds) {
            findViewById(buttonId).setOnClickListener(view -> switchDataSet(((Button) view).getText().toString()));
        }

        applyInsets(findViewById(R.id.map_container));
    }

    /**
     * Switches the currently displayed dataset to the one specified by the provided label.
     * <p>
     * This method attempts to find a DataSet object associated with the given label.
     * If a matching DataSet is found, it updates the map's feature layer to display the
     * data from that dataset. It also applies styling to the new dataset layer and centers
     * the map on the dataset's specified location.
     * <p>
     * If no matching DataSet is found, a toast message is displayed indicating the failure.
     * <p>
     * @param label The label of the dataset to switch to. This label should correspond to a
     *              dataset that has been previously added or loaded.
     */
    private void switchDataSet(String label) {
        DataSet dataSet = findDataSetByLabel(label);
        if (dataSet == null) {
            Toast.makeText(this, "Failed to find dataset" + label, Toast.LENGTH_SHORT).show();
        } else {
            datasetLayer = map.getFeatureLayer(
                    new FeatureLayerOptions.Builder()
                            .featureType(FeatureType.DATASET)
                            .datasetId(dataSet.datasetId())
                            .build()
            );
            dataSet.callback.styleDatasetLayer();
            centerMapOnLocation(dataSet.location(), dataSet.zoomLevel());
        }
    }

    @Override
    public void onMapReady(@NonNull GoogleMap googleMap) {
        this.map = googleMap;

        MapCapabilities capabilities = map.getMapCapabilities();
        Log.d(TAG, "Data-driven Styling is available: " + capabilities.isDataDrivenStylingAvailable());
        if (!capabilities.isDataDrivenStylingAvailable()) {
            Toast.makeText(
                    this,
                    "Data-driven Styling is not available.  See README.md for instructions.",
                    Toast.LENGTH_LONG
            ).show();
        }

        // Switch to the default dataset which must happen before adding the feature click listener
        switchDataSet("Boulder");

        datasetLayer.addOnFeatureClickListener(this);
    }

    private void styleNYCDatasetLayer() {
        FeatureLayer.StyleFactory styleFactory = feature -> {
            int fillColor = Color.GREEN;
            int strokeColor = Color.YELLOW;
            float pointRadius = 12F;

            if (feature instanceof DatasetFeature) {
                Map<String, String> furColors = ((DatasetFeature) feature).getDatasetAttributes();
                String furColor = furColors.get("Color");

                if (furColor != null) {
                    switch (furColor) {
                        case "Black+":
                            fillColor = Color.BLACK;
                            strokeColor = Color.BLACK;
                            break;
                        case "Cinnamon+":
                            fillColor = 0xFF8B0000; // dark red color
                            strokeColor = 0xFF8B0000;
                            break;
                        case "Cinnamon+Gray":
                            fillColor = 0xFF8B0000; // dark red color
                            strokeColor = 0xFF8B0000;
                            pointRadius = 10F;
                            break;
                        case "Cinnamon+White":
                            fillColor = 0xFF8B0000; // dark red color
                            strokeColor = Color.WHITE;
                            pointRadius = 10F;
                            break;
                        case "Gray+":
                            fillColor = Color.GRAY;
                            break;
                        case "Gray+Cinnamon":
                            fillColor = Color.GRAY;
                            strokeColor = 0xFF8B0000; // dark red color
                            pointRadius = 10F;
                            break;
                        case "Gray+Cinnamon, White":
                            fillColor = Color.LTGRAY;
                            strokeColor = 0xFF8B0000; // dark red color
                            pointRadius = 10F;
                            break;
                        case "Gray+White":
                            fillColor = Color.GRAY;
                            strokeColor = Color.WHITE;
                            pointRadius = 10F;
                            break;
                    }
                }

                return new FeatureStyle.Builder()
                        .fillColor(fillColor)
                        .strokeColor(strokeColor)
                        .pointRadius(pointRadius)
                        .build();
            }
            return null;
        };

        if (datasetLayer != null) {
            datasetLayer.setFeatureStyle(styleFactory);
        }
    }

    private void styleKyotoDatasetsLayer() {
        // Create the style factory function.
        FeatureLayer.StyleFactory styleFactory = feature -> {
            // Check if the feature is an instance of DatasetFeature.
            if (feature instanceof DatasetFeature datasetFeature) {
                // Determine the value of the typecategory attribute.
                Map<String, String> typeCategories = datasetFeature.getDatasetAttributes();
                String typeCategory = typeCategories.get("type");

                // Set default colors to green.
                int fillColor;
                int strokeColor;

                if ("temple".equals(typeCategory)) {
                    // Color temples areas blue.
                    fillColor = Color.BLUE;
                    strokeColor = Color.BLUE;
                } else {
                    // Color all other areas green.
                    fillColor = Color.GREEN;
                    strokeColor = Color.GREEN;
                }

                return new FeatureStyle.Builder()
                        .fillColor(fillColor)
                        .strokeColor(strokeColor)
                        .strokeWidth(2F)
                        .build();
            }
            return null;
        };

        // Apply the style factory function to the feature layer.
        if (datasetLayer != null) {
            datasetLayer.setFeatureStyle(styleFactory);
        }
    }

    private void styleBoulderDatasetLayer() {
        final int EASY = Color.GREEN;
        final int MODERATE = Color.BLUE;
        final int DIFFICULT = Color.RED;

        // Create the style factory function.
        FeatureLayer.StyleFactory styleFactory = feature -> {
            // Set default colors to yellow and point radius to 8.
            int fillColor;
            int strokeColor;
            float pointRadius = 8F;
            float strokeWidth = 3F;

            // Check if the feature is an instance of DatasetFeature.
            if (feature instanceof DatasetFeature datasetFeature) {
                Map<String, String> attributes = datasetFeature.getDatasetAttributes();
                String difficulty = attributes.get("OSMPTrailsOSMPDIFFICULTY");
                String name = attributes.get("OSMPTrailsOSMPTRAILNAME");
                String dogsAllowed = attributes.get("OSMPTrailsOSMPDOGREGGEN");

                if ("Easy".equals(difficulty)) {
                    fillColor = EASY;
                } else if ("Moderate".equals(difficulty)) {
                    fillColor = MODERATE;
                } else if ("Difficult".equals(difficulty)) {
                    fillColor = DIFFICULT;
                } else {
                    Log.w(TAG, name + " -> Unknown difficulty: " + difficulty);
                    fillColor = Color.MAGENTA;
                }

                if ("No Dogs".equals(dogsAllowed)) {
                    fillColor = ColorUtils.setAlphaComponent(fillColor, 66);
                    strokeWidth = 5f;
                } else if ("LVS".equals(dogsAllowed)) {
                    // No change needed
                } else if ("LR".equals(dogsAllowed) || "RV".equals(dogsAllowed)) {
                    fillColor = ColorUtils.setAlphaComponent(fillColor, 75);
                } else {
                    Log.w(TAG, name + " -> Unknown dogs reg: " + dogsAllowed);
                }

                strokeColor = fillColor;

                return new FeatureStyle.Builder()
                        .fillColor(fillColor)
                        .strokeColor(strokeColor)
                        .pointRadius(pointRadius)
                        .strokeWidth(strokeWidth)
                        .build();
            }
            return null;
        };

        // Apply the style factory function to the feature layer.
        if (datasetLayer != null) {
            datasetLayer.setFeatureStyle(styleFactory);
        }
    }


    private void centerMapOnLocation(LatLng location, float zoomLevel) {
        map.moveCamera(CameraUpdateFactory.newLatLngZoom(location, zoomLevel));
    }

    @Override
    public void onFeatureClick(FeatureClickEvent event) {
        List<Feature> clickFeatures = event.getFeatures();
        lastGlobalId = null;
        if (clickFeatures.get(0) instanceof DatasetFeature) {
            lastGlobalId = ((DatasetFeature) clickFeatures.get(0)).getDatasetAttributes().get("globalid");
            styleDatasetsLayerClickEvent();
        }
    }

    private void styleDatasetsLayerClickEvent() {
        FeatureLayer.StyleFactory styleFactory = feature -> {
            if (feature instanceof DatasetFeature) {
                Map<String, String> globalIDs = ((DatasetFeature) feature).getDatasetAttributes();
                String globalID = globalIDs.get("globalid");
                int fillColor = Color.GREEN;
                int strokeColor = Color.GREEN;

                if (globalID != null && globalID.equals(lastGlobalId)) {
                    fillColor = Color.BLUE;
                    strokeColor = Color.BLUE;
                }

                return new FeatureStyle.Builder()
                        .fillColor(fillColor)
                        .strokeColor(strokeColor)
                        .build();
            }
            return null;
        };

        if (datasetLayer != null) {
            datasetLayer.setFeatureStyle(styleFactory);
        }
    }
}
// [END maps_android_data_driven_styling_datasets]

File: ApiDemos/project/kotlin-app/src/main/java/com/example/kotlindemos/DataDrivenDatasetStylingActivity.kt
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.example.kotlindemos

import android.graphics.Color
import android.os.Build

import com.google.android.gms.maps.OnMapReadyCallback
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.WindowInsets
import android.widget.Button
import android.widget.LinearLayout
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.annotation.ColorInt
import androidx.core.graphics.ColorUtils
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.model.DatasetFeature
import com.google.android.gms.maps.model.Feature
import com.google.android.gms.maps.model.FeatureClickEvent
import com.google.android.gms.maps.model.FeatureLayer
import com.google.android.gms.maps.model.FeatureLayerOptions
import com.google.android.gms.maps.model.FeatureStyle
import com.google.android.gms.maps.model.FeatureType
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MapCapabilities
import androidx.core.view.WindowCompat
import com.google.android.gms.maps.GoogleMapOptions


private val TAG = DataDrivenDatasetStylingActivity::class.java.name

/**
 * This sample showcases how to use the Data-driven styling for datasets. For more information
 * on how the Data-driven styling for boundaries work, check out the following link:
 * https://developers.google.com/maps/documentation/android-sdk/dds-datasets/overview
 */
class DataDrivenDatasetStylingActivity : SamplesBaseActivity(), OnMapReadyCallback, FeatureLayer.OnFeatureClickListener {
    private lateinit var mapContainer: ViewGroup

    private lateinit var map: GoogleMap
    private val zoomLevel = 13.5f

    private var datasetLayer: FeatureLayer? = null

    // The global id of the clicked dataset feature.
    private var lastGlobalId: String? = null

    private data class DataSet(
        val datasetId: String,
        val location: LatLng,
        val callback: DataDrivenDatasetStylingActivity.() -> Unit
    )

    private val dataSets = mutableMapOf<String, DataSet>()

    private lateinit var buttonLayout: LinearLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)

        val mapId = (application as ApiDemoApplication).mapId

        // --- Map ID Check ---
        if (mapId == null) {
            finish()
            return // Exit early if no valid Map ID
        }

        if (dataSets.isEmpty()) {
            with(dataSets) {
                put(getString(com.example.common_ui.R.string.boulder), DataSet(BuildConfig.BOULDER_DATASET_ID, LatLng(40.0150, -105.2705)) { styleBoulderDataset() })
                put(getString(com.example.common_ui.R.string.new_york), DataSet(BuildConfig.NEW_YORK_DATASET_ID, LatLng(40.786244, -73.962684)) { styleNYCDataset() })
                put(getString(com.example.common_ui.R.string.kyoto), DataSet(BuildConfig.KYOTO_DATASET_ID, LatLng(35.005081, 135.764385)) { styleKyotoDataset() })
            }
        }

        setContentView(com.example.common_ui.R.layout.data_driven_styling_demo)

        mapContainer = findViewById(com.example.common_ui.R.id.map_container)

        // --- Programmatically create and add the map fragment ---
        // 1. Create GoogleMapOptions
        val mapOptions = GoogleMapOptions().apply {
            // 2. Set the mapId using your BuildConfig field
            mapId(mapId)
        }

        // 3. Create SupportMapFragment instance with options
        val mapFragment = SupportMapFragment.newInstance(mapOptions)

        // 4. Add the fragment to your FrameLayout container
        supportFragmentManager.beginTransaction()
            .replace(com.example.common_ui.R.id.map_fragment_container, mapFragment) // Use the container ID from XML
            .commit()
        // --- End of programmatic creation ---

        mapFragment.getMapAsync(this)

        // Set the click listener for each of the buttons
        listOf(com.example.common_ui.R.id.button_kyoto, com.example.common_ui.R.id.button_ny, com.example.common_ui.R.id.button_boulder).forEach { viewId ->
            findViewById<Button>(viewId).setOnClickListener { view ->
                switchToDataset((view as Button).text.toString())
            }
        }

        buttonLayout = findViewById<View>(com.example.common_ui.R.id.button_kyoto).parent as LinearLayout

        handleCutout()
        applyInsets(findViewById<View?>(com.example.common_ui.R.id.map_container))
    }

    private fun handleCutout() {
        WindowCompat.setDecorFitsSystemWindows(window, false)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.decorView.setOnApplyWindowInsetsListener { _, windowInsets ->
                val insets = windowInsets.getInsets(WindowInsets.Type.systemBars())
                val topInset = insets.top
                mapContainer.setPadding(0, topInset, 0, 0)
                windowInsets
            }
        } else {
            window.decorView.setOnApplyWindowInsetsListener { view, windowInsets ->
                val topInset = windowInsets.systemWindowInsetTop
                mapContainer.setPadding(0, topInset, 0, 0)
                windowInsets
            }
        }
    }

    /**
     * Switches the currently displayed dataset on the map to the one identified by the given label.
     *
     * This function retrieves a DataSetInfo object from the `dataSets` map using the provided `label`.
     * If a dataset with the given label exists, it does the following:
     *   1. Creates a new FeatureLayer for the specified dataset.
     *   2. Sets the `datasetLayer` property to the newly created FeatureLayer.
     *   3. Executes the callback function associated with the dataset, passing the current activity instance (this).
     *   4. Centers the map on the location associated with the dataset.
     * If no dataset with the given label is found, it displays a Toast message indicating an unknown dataset.
     *
     * @param label The label identifying the dataset to switch to. This label should correspond to a key in the `dataSets` map.
     * @throws IllegalStateException if `map` is not initialized.
     */
    private fun switchToDataset(label: String) {
        dataSets[label]?.let { dataSet ->
            datasetLayer = map.getFeatureLayer(
                with(FeatureLayerOptions.Builder()) {
                    featureType(FeatureType.DATASET)
                    // Specify the dataset ID.
                    datasetId(dataSet.datasetId)
                }.build()
            )
            dataSet.callback(this)
            centerMapOnLocation(dataSet.location)
        } ?: run {
            Toast.makeText(this, "Unknown dataset: $label", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        map = googleMap

        val capabilities: MapCapabilities = map.mapCapabilities
        println("Data-driven Styling is available: " + capabilities.isDataDrivenStylingAvailable)

        switchToDataset("Boulder")

        // Register the click event handler for the Datasets layer.
        datasetLayer?.addOnFeatureClickListener(this)
    }

    private fun styleNYCDataset() {
        data class Style(
            @ColorInt val fillColor: Int,
            @ColorInt val strokeColor: Int,
            val pointRadius: Float,
        )

        val largePointRadius = 8F
        val smallPointRadius = 6F
        val darkRedBrown = resources.getColor(R.color.darkRedBrown)

        val styleFactory = FeatureLayer.StyleFactory { feature: Feature ->
            if (feature is DatasetFeature) {
                val furColors: MutableMap<String, String> = feature.datasetAttributes
                val furColor = furColors["Color"]

                val style = when (furColor) {
                    "Black+" -> Style(Color.BLACK, Color.BLACK, largePointRadius)
                    "Cinnamon+" -> Style(darkRedBrown, darkRedBrown, largePointRadius)
                    "Cinnamon+Gray" -> Style(darkRedBrown, darkRedBrown, smallPointRadius)
                    "Cinnamon+White" -> Style(darkRedBrown, Color.WHITE, smallPointRadius)
                    "Gray+" -> Style(Color.GRAY, Color.YELLOW, largePointRadius) // Default stroke
                    "Gray+Cinnamon" -> Style(Color.GRAY, darkRedBrown, smallPointRadius)
                    "Gray+Cinnamon, White" -> Style(Color.LTGRAY, darkRedBrown, smallPointRadius)
                    "Gray+White" -> Style(Color.GRAY, Color.WHITE, smallPointRadius)
                    else -> Style(Color.GREEN, Color.YELLOW, largePointRadius) // Default style if furColor is null or doesn't match
                }

                return@StyleFactory FeatureStyle.Builder()
                    .fillColor(style.fillColor)
                    .strokeColor(style.strokeColor)
                    .pointRadius(style.pointRadius)
                    .build()
            }
            return@StyleFactory null
        }

        // Apply the style factory function to the feature layer.
        datasetLayer?.featureStyle = styleFactory
    }

    private fun styleKyotoDataset() {

        // Create the style factory function.
        val styleFactory = FeatureLayer.StyleFactory { feature: Feature ->

            // Check if the feature is an instance of DatasetFeature.
            if (feature is DatasetFeature) {
                // Determine the value of the typecategory attribute.
                val typeCategories: MutableMap<String, String> = feature.datasetAttributes
                val typeCategory = typeCategories["type"]
                // Set default colors to green.
                val fillColor: Int
                val strokeColor: Int
                when (typeCategory) {
                    "temple" -> {
                        // Color temples areas blue.
                        fillColor = Color.BLUE
                        strokeColor = Color.BLUE
                    }

                    else -> {
                        // Color all other areas green.
                        fillColor = Color.GREEN
                        strokeColor = Color.GREEN
                    }
                }
                return@StyleFactory FeatureStyle.Builder()
                    .fillColor(fillColor)
                    .strokeColor(strokeColor)
                    .strokeWidth(2F)
                    .build()
            }
            return@StyleFactory null
        }

        // Apply the style factory function to the feature layer.
        datasetLayer?.featureStyle = styleFactory
    }

    private fun styleBoulderDataset() {
        val EASY = Color.GREEN
        val MODERATE = Color.BLUE
        val DIFFICULT = Color.RED

        // Create the style factory function.
        val styleFactory = FeatureLayer.StyleFactory { feature: Feature ->

            // Set default colors to to yellow and point radius to 8.
            var fillColor: Int
            var strokeColor: Int
            val pointRadius = 8F
            var strokeWidth = 3F
            // Check if the feature is an instance of DatasetFeature.
            if (feature is DatasetFeature) {

                val attributes: MutableMap<String, String> = feature.datasetAttributes
                val difficulty = attributes["OSMPTrailsOSMPDIFFICULTY"]
                val name = attributes["OSMPTrailsOSMPTRAILNAME"]
                val dogsAllowed = attributes["OSMPTrailsOSMPDOGREGGEN"]

                when (difficulty) {
                    "Easy" -> {
                        fillColor = EASY
                    }
                    "Moderate" -> {
                        fillColor = MODERATE
                    }
                    "Difficult" -> {
                        fillColor = DIFFICULT
                    }
                    else -> {
                        Log.w(TAG, "$name -> Unknown difficulty: $difficulty")
                        fillColor = Color.MAGENTA
                    }
                }

                when (dogsAllowed) {
                    "No Dogs" -> {
                        fillColor = ColorUtils.setAlphaComponent(fillColor, 66)
                        strokeWidth = 5f
                    }
                    "LVS" -> {
                    }
                    "LR", "RV" -> {
                        fillColor = ColorUtils.setAlphaComponent(fillColor, 75)
                    }
                    else -> {
                        Log.w(TAG, "$name -> Unknown dogs reg: $dogsAllowed")
                    }
                }

                strokeColor = fillColor

                return@StyleFactory FeatureStyle.Builder()
                    .fillColor(fillColor)
                    .strokeColor(strokeColor)
                    .pointRadius(pointRadius)
                    .strokeWidth(strokeWidth)
                    .build()
            }
            return@StyleFactory null
        }

        // Apply the style factory function to the feature layer.
        datasetLayer?.featureStyle = styleFactory
    }


    private fun centerMapOnLocation(location: LatLng) {
        map.moveCamera(CameraUpdateFactory.newLatLngZoom(location, zoomLevel))
    }

    // Define the click event handler to set lastGlobalId to globalid of selected feature.
    override fun onFeatureClick(event: FeatureClickEvent) {
        // Get the dataset feature affected by the click.
        val clickFeatures: MutableList<Feature> = event.features
        lastGlobalId = null
        if (clickFeatures[0] is DatasetFeature) {
            lastGlobalId = ((clickFeatures[0] as DatasetFeature).datasetAttributes["globalid"])
            // Remember to reset the Style Factory.
            styleDatasetsLayerClickEvent()
        }
    }

    // Set fill and border for all features.
    private fun styleDatasetsLayerClickEvent() {
        // Create the style factory function.
        val styleFactory = FeatureLayer.StyleFactory { feature: Feature ->

            // Check if the feature is an instance of DatasetFeature.
            if (feature is DatasetFeature) {
                val globalIDs: MutableMap<String, String> = feature.datasetAttributes
                // Determine globalid attribute.
                val globalID = globalIDs["globalid"]
                // Set default colors to to green.
                var fillColor = Color.GREEN
                var strokeColor = Color.GREEN
                if (globalID == lastGlobalId) {
                    // Color selected area blue.
                    fillColor = Color.BLUE
                    strokeColor = Color.BLUE
                }
                return@StyleFactory FeatureStyle.Builder()
                    .fillColor(fillColor)
                    .strokeColor(strokeColor)
                    .build()
            }
            return@StyleFactory null
        }

        // Apply the style factory function to the dataset feature layer.
        datasetLayer?.featureStyle = styleFactory
    }
}
